File: tools/memory_cotravel.py
#!/usr/bin/env python3
r"""memory_cotravel.py -- carry a session's slug-scoped agent-memory store along
with its transcript when the resume layer re-homes it onto a healthy account.

THE SEAM THIS CLOSES. Each ``~/.claude*`` account dir holds, PER PROJECT SLUG, two
parallel stores: the conversation transcript ``projects/<slug>/<sid>.jsonl`` AND the
agent auto-memory store ``projects/<slug>/memory/*.md`` (the same store
:mod:`sync_memory` mirrors home<->repo, resolved by ``default_home_memory`` as
``$CLAUDE_CONFIG_DIR/projects/<slug>/memory``). The switcher's resume path re-homes a
session from a throttled account A onto a healthy account B by copying the transcript
(``fleet_resume_watchdog.rehome_transcript``) and pinning ``CLAUDE_CONFIG_DIR=B``. But
the memory store did NOT travel: the resumed session reads/writes memory under B's
slug store, which may be empty or stale, while everything it learned lives under A.
The session lands memory-amnesic. This module is the missing co-travel.

DESIGNED AS AN EXPERIMENTATION SURFACE, not a frozen one-shot:

  * The per-file MERGE decision is a NAMED, PLUGGABLE strategy (the ``STRATEGIES``
    registry). Adding a new way to reconcile a conflict is ~3 lines + a unit test;
    NO core re-home code changes. Default ``additive`` (never clobber a dest memory).

  * The whole co-travel runs behind a SHADOW gate (``FAK_MEMORY_COTRAVEL``), mirroring
    :mod:`switcher_shadow`: in ``shadow`` it COMPUTES every per-file decision and
    appends it to a host-local, off-repo, append-only ledger but copies NOTHING -- so a
    real re-home can be OBSERVED (how often memory would move, how many conflicts each
    strategy hits, whether ``additive``'s skips ever drop a needed memory) with zero
    risk. ``live`` actually copies. ``off`` is the original transcript-only behavior.

Gate / strategy (env, overridable per-call):
    FAK_MEMORY_COTRAVEL = shadow (default) | live | off
    FAK_MEMORY_MERGE    = additive (default) | source_wins | newest_mtime

Ledger (host-local, off-repo, append-only, atomic, size-capped) -- same discipline as
switcher_shadow's ledger: ``~/.claude/fak-memory-cotravel-ledger.jsonl``.
"""
from __future__ import annotations

import glob
import json
import os
import shutil
import tempfile
from datetime import datetime, timezone

# --------------------------------------------------------------------------- #
# Gate + ledger location (env-overridable; read at call time so tests can patch).
# --------------------------------------------------------------------------- #
HOME = os.environ.get("FLEET_USER_HOME", os.path.expanduser("~"))
LEDGER_PATH = os.environ.get(
    "FAK_MEMORY_COTRAVEL_LEDGER",
    os.path.join(HOME, ".claude", "fak-memory-cotravel-ledger.jsonl"),
)
LEDGER_CAP_BYTES = int(os.environ.get("FAK_MEMORY_COTRAVEL_LEDGER_CAP", str(8 * 1024 * 1024)))

# The default gate is SHADOW: the feature observes real re-homes and copies nothing
# until an operator flips it to ``live``. This is the prove-before-trust default the
# stability doctrine wants -- the amnesia gap is measured immediately and closed on the
# flip, never opened blind.
_DEFAULT_GATE = "shadow"
_DEFAULT_STRATEGY = "additive"
_VALID_GATES = ("shadow", "live", "off")


def gate() -> str:
    g = (os.environ.get("FAK_MEMORY_COTRAVEL") or _DEFAULT_GATE).strip().lower()
    return g if g in _VALID_GATES else _DEFAULT_GATE


def strategy_name() -> str:
    s = (os.environ.get("FAK_MEMORY_MERGE") or _DEFAULT_STRATEGY).strip().lower()
    return s if s in STRATEGIES else _DEFAULT_STRATEGY


# --------------------------------------------------------------------------- #
# Pluggable per-file merge strategies: (src_path, dst_path) -> "copy" | "skip".
# Pure functions of the two file paths -- no I/O beyond stat/read for the compare,
# trivially unit-testable, and NONE of them ever prune a dest-only file (that policy
# lives in the caller, which only ever iterates SOURCE files).
# --------------------------------------------------------------------------- #
def _differ(src: str, dst: str) -> bool:
    """True when dst is absent or its bytes differ from src. Mirrors
    sync_memory.differ -- the same byte-compare the repo<->home mirror uses."""
    if not os.path.exists(dst):
        return True
    try:
        with open(src, "rb") as a, open(dst, "rb") as b:
            return a.read() != b.read()
    except OSError:
        return True


def _additive(src: str, dst: str) -> str:
    """Never clobber. Copy only when the dest file is MISSING; a same-name dest with
    different bytes is kept (a peer / newer memory under the target must survive). The
    safe, regression-proof default."""
    return "copy" if not os.path.exists(dst) else "skip"


def _source_wins(src: str, dst: str) -> str:
    """Source authoritative: copy whenever the bytes differ (overwrite a conflicting
    dest). Guarantees the resumed session sees exactly A's memory -- at the cost of
    clobbering a newer/peer memory already under B."""
    return "copy" if _differ(src, dst) else "skip"


def _newest_mtime(src: str, dst: str) -> str:
    """Keep whichever file is newer by mtime. FLAGGED UNRELIABLE: shutil.copy2 preserves
    the source mtime, so a re-homed copy ties its origin -- the same mtime trap
    resume_resolver documents for transcript owner selection. Present for experiments,
    NOT the default."""
    if not os.path.exists(dst):
        return "copy"
    try:
        return "copy" if os.path.getmtime(src) > os.path.getmtime(dst) else "skip"
    except OSError:
        return "skip"


STRATEGIES = {
    "additive": _additive,
    "source_wins": _source_wins,
    "newest_mtime": _newest_mtime,
}


# --------------------------------------------------------------------------- #
# The co-travel itself -- decide per *.md, then act per the gate.
# --------------------------------------------------------------------------- #
def _now_iso() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def plan_one_dir(src_mem: str, dst_mem: str, decide) -> list[dict]:
    """Per-file decisions for carrying *.md from src_mem -> dst_mem under ``decide``.

    Iterates SOURCE files only (so a dest-only memory is never even considered for a
    prune -- co-travel is strictly additive in surface). Returns a list of
    {name, action, dst_exists} records; pure (no copy)."""
    plan: list[dict] = []
    if not os.path.isdir(src_mem):
        return plan
    for src in sorted(glob.glob(os.path.join(src_mem, "*.md"))):
        name = os.path.basename(src)
        dst = os.path.join(dst_mem, name)
        # Never act on a self-copy (mirroring within the owner account resolves dst==src).
        if os.path.abspath(dst) == os.path.abspath(src):
            plan.append({"name": name, "action": "skip", "dst_exists": True})
            continue
        plan.append({"name": name, "action": decide(src, dst),
                     "dst_exists": os.path.exists(dst)})
    return plan


def cotravel_memory(src_cfg: str, dst_cfg: str, slug: str, sid: str,
                    *, dst_slug: str | None = None,
                    gate_value: str | None = None,
                    strategy: str | None = None) -> dict:
    """Carry ``projects/<slug>/memory/*.md`` from src_cfg -> dst_cfg for one re-home.

    The SOURCE memory is always read from the session's owner slug (``slug``). The
    DESTINATION defaults to that same slug, but ``dst_slug`` lets the caller land it
    under a DIFFERENT slug -- the cross-directory resume case: a session born under
    ``C--work-fak`` whose memory lives under that slug must be carried into the LAUNCHING
    cwd's slug (e.g. ``C--work-slack-helpers``) where ``claude --resume`` will look from
    that folder. So src memory = owner slug, dst memory = dst_slug (or owner slug).

    Honors the gate: ``off`` -> no-op; ``shadow`` -> decide + ledger, copy nothing;
    ``live`` -> decide + actually copy the ``copy`` files. Returns a record describing
    what happened (also the ledger row in shadow/live). Fail-open by construction: any
    single file copy that raises (e.g. a Windows mandatory lock) is swallowed, never
    crashing the resolver -- the launcher's contract."""
    g = (gate_value or gate())
    strat = (strategy or strategy_name())
    decide = STRATEGIES.get(strat, _additive)
    dst_slug = dst_slug or slug
    src_mem = os.path.join(src_cfg, "projects", slug, "memory")
    dst_mem = os.path.join(dst_cfg, "projects", dst_slug, "memory")

    rec = {
        "ts": _now_iso(), "session": sid, "slug": slug, "dst_slug": dst_slug,
        "gate": g, "strategy": strat,
        "src_memory": src_mem, "dst_memory": dst_mem,
        "src_has_memory": os.path.isdir(src_mem),
    }
    if g == "off":
        rec.update({"plan": [], "copied": [], "skipped": [], "note": "gate=off (no-op)"})
        return rec

    plan = plan_one_dir(src_mem, dst_mem, decide)
    rec["plan"] = plan
    to_copy = [p["name"] for p in plan if p["action"] == "copy"]
    rec["skipped"] = [p["name"] for p in plan if p["action"] != "copy"]

    if g == "shadow":
        rec["copied"] = []
        rec["would_copy"] = to_copy
        _append_ledger(rec)
        return rec

    # g == "live": actually copy the chosen files (fail-open per file).
    copied: list[str] = []
    if to_copy:
        try:
            os.makedirs(dst_mem, exist_ok=True)
        except OSError:
            pass
    for name in to_copy:
        try:
            shutil.copy2(os.path.join(src_mem, name), os.path.join(dst_mem, name))
            copied.append(name)
        except OSError:
            continue  # locked / vanished -- skip this one, keep going
    rec["copied"] = copied
    _append_ledger(rec)
    return rec


# --------------------------------------------------------------------------- #
# Host-local ledger (off-repo, append-only, atomic, size-capped) -- switcher_shadow shape.
# --------------------------------------------------------------------------- #
def _ledger_path() -> str:
    return os.environ.get("FAK_MEMORY_COTRAVEL_LEDGER", LEDGER_PATH)


def _rotate_if_needed(path: str) -> None:
    try:
        if os.path.getsize(path) > LEDGER_CAP_BYTES:
            os.replace(path, path + ".1")  # keep one prior generation; bounded
    except OSError:
        pass


def _append_ledger(row: dict) -> None:
    """Append one co-travel row. Atomic temp-file + os.replace rewrite (O_APPEND is not
    safe across concurrent re-homes on Windows). Best-effort -- a ledger write must never
    crash the re-home."""
    path = _ledger_path()
    try:
        _rotate_if_needed(path)
        os.makedirs(os.path.dirname(path) or ".", exist_ok=True)
        existing = []
        if os.path.exists(path):
            with open(path, encoding="utf-8") as f:
                existing = [ln for ln in f.read().splitlines() if ln.strip()]
        fd, tmp = tempfile.mkstemp(dir=os.path.dirname(path) or ".", suffix=".tmp")
        with os.fdopen(fd, "w", encoding="utf-8") as f:
            for ln in existing:
                f.write(ln + "\n")
            f.write(json.dumps(row) + "\n")
        os.replace(tmp, path)
    except OSError:
        pass


def read_ledger() -> list[dict]:
    rows: list[dict] = []
    try:
        with open(_ledger_path(), encoding="utf-8") as f:
            for ln in f:
                ln = ln.strip()
                if ln:
                    try:
                        rows.append(json.loads(ln))
                    except ValueError:
                        continue
    except OSError:
        pass
    return rows

File: internal/memorycotravel/memorycotravel.go
package memorycotravel

import (
	"bufio"
	"bytes"
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/anthony-chaudhary/fak/internal/rolloutmode"
)

// Invariant: memory cotravel evaluation is fail-closed and bounded.
// Guard: rollout gate defaults to shadow mode to prevent unintended disk writes across workspace boundaries.
// Contract: memory files are transferred only when allowed by the active strategy and gate.

const (
	// DefaultGate is the safe rollout mode used when no environment or option override is provided.
	DefaultGate = string(rolloutmode.Shadow)
	// DefaultStrategy is the fallback merge policy that preserves existing destination files.
	DefaultStrategy = "additive"
	// DefaultLedgerCap specifies the maximum byte threshold before the JSONL audit ledger undergoes single-generation rotation.
	DefaultLedgerCap = 8 * 1024 * 1024
)

// Decision computes the planned transfer action ("copy" or "skip") for a given source and destination file.
type Decision func(src, dst string) string

// PlanItem records the prospective transfer action and destination file presence for a single memory entry.
type PlanItem struct {
	Name      string `json:"name"`
	Action    string `json:"action"`
	DstExists bool   `json:"dst_exists"`
}

// Record captures an auditable memory synchronization execution including gate, strategy, plan, and outcome.
type Record struct {
	Timestamp    string     `json:"ts"`
	Session      string     `json:"session"`
	Slug         string     `json:"slug"`
	DstSlug      string     `json:"dst_slug"`
	Gate         string     `json:"gate"`
	Strategy     string     `json:"strategy"`
	SrcMemory    string     `json:"src_memory"`
	DstMemory    string     `json:"dst_memory"`
	SrcHasMemory bool       `json:"src_has_memory"`
	Plan         []PlanItem `json:"plan"`
	Copied       []string   `json:"copied"`
	Skipped      []string   `json:"skipped"`
	WouldCopy    []string   `json:"would_copy,omitempty"`
	Note         string     `json:"note,omitempty"`
}

// Strategies maps named resolution policies to their corresponding Decision implementation functions.
var Strategies = map[string]Decision{
	"additive":     Additive,
	"source_wins":  SourceWins,
	"newest_mtime": NewestMtime,
}

var memoryCotravelModes = []rolloutmode.Mode{rolloutmode.Off, rolloutmode.Shadow, rolloutmode.On}

func parseGate(raw string) rolloutmode.Mode {
	raw = strings.ToLower(strings.TrimSpace(raw))
	// Accept the retired applied-rung spelling at the environment boundary while
	// rendering and recording only the shared canonical `on` rung (#6090).
	if raw == "live" {
		raw = rolloutmode.On.String()
	}
	m, _ := rolloutmode.ParseIn(raw, memoryCotravelModes, rolloutmode.Shadow)
	return m
}

// Gate returns the active rollout gate string parsed from FAK_MEMORY_COTRAVEL.
func Gate() string { return parseGate(os.Getenv("FAK_MEMORY_COTRAVEL")).String() }

// StrategyName returns the configured resolution policy name parsed from FAK_MEMORY_MERGE.
func StrategyName() string {
	s := strings.ToLower(strings.TrimSpace(os.Getenv("FAK_MEMORY_MERGE")))
	if _, ok := Strategies[s]; ok {
		return s
	}
	return DefaultStrategy
}

// Differ compares two files on disk and returns true if their byte contents differ or either read fails.
func Differ(src, dst string) bool {
	a, err := os.ReadFile(src)
	if err != nil {
		return true
	}
	b, err := os.ReadFile(dst)
	if err != nil {
		return true
	}
	return !bytes.Equal(a, b)
}

// Additive implements a non-destructive conflict policy that copies only when the destination file is missing.
func Additive(_src, dst string) string {
	if _, err := os.Stat(dst); err == nil {
		return "skip"
	}
	return "copy"
}

// SourceWins implements a source-preferred conflict policy that copies when the source and destination differ.
func SourceWins(src, dst string) string {
	if Differ(src, dst) {
		return "copy"
	}
	return "skip"
}

// NewestMtime implements a timestamp-based conflict policy that copies when the source file has a newer modification time.
func NewestMtime(src, dst string) string {
	dstInfo, err := os.Stat(dst)
	if err != nil {
		return "copy"
	}
	srcInfo, err := os.Stat(src)
	if err != nil {
		return "skip"
	}
	if srcInfo.ModTime().After(dstInfo.ModTime()) {
		return "copy"
	}
	return "skip"
}

// PlanOneDir inspects the source memory directory and evaluates decision rules against corresponding destination paths.
func PlanOneDir(srcMem, dstMem string, decide Decision) []PlanItem {
	info, err := os.Stat(srcMem)
	if err != nil || !info.IsDir() {
		return nil
	}
	paths, _ := filepath.Glob(filepath.Join(srcMem, "*.md"))
	sort.Strings(paths)
	plan := make([]PlanItem, 0, len(paths))
	for _, src := range paths {
		name := filepath.Base(src)
		dst := filepath.Join(dstMem, name)
		if sameAbs(src, dst) {
			plan = append(plan, PlanItem{Name: name, Action: "skip", DstExists: true})
			continue
		}
		_, statErr := os.Stat(dst)
		plan = append(plan, PlanItem{Name: name, Action: decide(src, dst), DstExists: statErr == nil})
	}
	return plan
}

// Options provides parameter overrides for memory cotravel execution including destination slug, gate, and strategy.
type Options struct {
	DstSlug  string
	Gate     string
	Strategy string
}

// CotravelMemory executes memory synchronization from source to destination configuration roots under the specified options.
func CotravelMemory(srcCfg, dstCfg, slug, sid string, opts Options) Record {
	g := parseGate(opts.Gate).String()
	if strings.TrimSpace(opts.Gate) == "" {
		g = Gate()
	}
	strat := opts.Strategy
	if strat == "" {
		strat = StrategyName()
	}
	decide := Strategies[strat]
	if decide == nil {
		strat = DefaultStrategy
		decide = Additive
	}
	dstSlug := opts.DstSlug
	if dstSlug == "" {
		dstSlug = slug
	}
	srcMem := filepath.Join(srcCfg, "projects", slug, "memory")
	dstMem := filepath.Join(dstCfg, "projects", dstSlug, "memory")
	rec := Record{
		Timestamp:    time.Now().UTC().Format("2006-01-02T15:04:05Z"),
		Session:      sid,
		Slug:         slug,
		DstSlug:      dstSlug,
		Gate:         g,
		Strategy:     strat,
		SrcMemory:    srcMem,
		DstMemory:    dstMem,
		SrcHasMemory: isDir(srcMem),
	}
	if g == rolloutmode.Off.String() {
		rec.Plan = []PlanItem{}
		rec.Copied = []string{}
		rec.Skipped = []string{}
		rec.Note = "gate=off (no-op)"
		return rec
	}
	rec.Plan = PlanOneDir(srcMem, dstMem, decide)
	for _, item := range rec.Plan {
		if item.Action == "copy" {
			rec.WouldCopy = append(rec.WouldCopy, item.Name)
		} else {
			rec.Skipped = append(rec.Skipped, item.Name)
		}
	}
	if g == rolloutmode.Shadow.String() {
		rec.Copied = []string{}
		AppendLedger(rec)
		return rec
	}
	if len(rec.WouldCopy) > 0 {
		_ = os.MkdirAll(dstMem, 0o755)
	}
	for _, name := range rec.WouldCopy {
		if err := copyFile(filepath.Join(srcMem, name), filepath.Join(dstMem, name)); err == nil {
			rec.Copied = append(rec.Copied, name)
		}
	}
	AppendLedger(rec)
	return rec
}

// LedgerPath resolves the audit log location using FAK_MEMORY_COTRAVEL_LEDGER or home directory defaults.
func LedgerPath() string {
	if p := os.Getenv("FAK_MEMORY_COTRAVEL_LEDGER"); strings.TrimSpace(p) != "" {
		return p
	}
	home := os.Getenv("FLEET_USER_HOME")
	if home == "" {
		if h, err := os.UserHomeDir(); err == nil {
			home = h
		}
	}
	return filepath.Join(home, ".claude", "fak-memory-cotravel-ledger.jsonl")
}

// AppendLedger records an auditable synchronization entry to the JSONL log file, performing rotation if needed.
func AppendLedger(row Record) {
	path := LedgerPath()
	if path == "" {
		return
	}
	_ = rotateIfNeeded(path)
	_ = os.MkdirAll(filepath.Dir(path), 0o755)
	var lines []string
	if f, err := os.Open(path); err == nil {
		sc := bufio.NewScanner(f)
		for sc.Scan() {
			if strings.TrimSpace(sc.Text()) != "" {
				lines = append(lines, sc.Text())
			}
		}
		_ = f.Close()
	}
	data, err := json.Marshal(row)
	if err != nil {
		return
	}
	tmp, err := os.CreateTemp(filepath.Dir(path), "*.tmp")
	if err != nil {
		return
	}
	tmpName := tmp.Name()
	for _, line := range lines {
		_, _ = tmp.WriteString(line + "\n")
	}
	_, _ = tmp.Write(append(data, '\n'))
	if err := tmp.Close(); err != nil {
		_ = os.Remove(tmpName)
		return
	}
	if err := os.Rename(tmpName, path); err != nil {
		_ = os.Remove(tmpName)
	}
}

// ReadLedger deserializes and returns all recorded synchronization entries from the active ledger path.
func ReadLedger() []map[string]any {
	path := LedgerPath()
	f, err := os.Open(path)
	if err != nil {
		return nil
	}
	defer f.Close()
	var rows []map[string]any
	sc := bufio.NewScanner(f)
	for sc.Scan() {
		line := strings.TrimSpace(sc.Text())
		if line == "" {
			continue
		}
		var row map[string]any
		if json.Unmarshal([]byte(line), &row) == nil {
			rows = append(rows, row)
		}
	}
	return rows
}

func rotateIfNeeded(path string) error {
	info, err := os.Stat(path)
	if err == nil && info.Size() > DefaultLedgerCap {
		return os.Rename(path, path+".1")
	}
	return nil
}

func sameAbs(a, b string) bool {
	aa, errA := filepath.Abs(a)
	bb, errB := filepath.Abs(b)
	return errA == nil && errB == nil && aa == bb
}

func isDir(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.IsDir()
}

func copyFile(src, dst string) error {
	data, err := os.ReadFile(src)
	if err != nil {
		return err
	}
	info, _ := os.Stat(src)
	if err := os.MkdirAll(filepath.Dir(dst), 0o755); err != nil {
		return err
	}
	if err := os.WriteFile(dst, data, 0o644); err != nil {
		return err
	}
	if info != nil {
		_ = os.Chtimes(dst, info.ModTime(), info.ModTime())
	}
	return nil
}
